---
title: "Project title"
output:
  html_document: default
  pdf_document: keep_md: true
---



# scRNA-seq analysis for construction of MCDMs, MO-MCDMs, and Connective Pathway Analysis

## Define data structure and input files

```{r echo=FALSE, include=FALSE}

dir.create("data")  
dir.create("data/IPA")  
dir.create("data/IPA/curation")  
dir.create("data/sorted_DGEs")  
dir.create("data/NicheNet_analysis_curated")

```

The expected data input are five csv.gz matrices, one per organ, with cells over columns and genes over rows.

```{r echo=FALSE}
### Read data from GSE and save it to data folder
filenames = c('data/Joint_UMI_expression_matrix.txt.gz',
              'data/Muscle_UMI_expression_matrix.txt.gz',
              'data/Lung_UMI_expression_matrix.txt.gz',
              'data/Skin_UMI_expression_matrix.txt.gz',
              'data/Spleen_UMI_expression_matrix.txt.gz')
```


## Quality assessment and full matrix construction

To ensure good quality data for downstream analyses, it is recommended to remove poor quality cells and genes from the analyses.
We performed these analyses for each organ separately. 
We define and keep the good quality cells as those having a minimum of 400 transcripts, 200 genes, and less than 20% mitochondrial genes. 
The good quality genes kept were defined as those being identified in at least 1% of the cells. 

Due to the risk of duplicates in the library resulting in two or more cells sharing a cell barcode, it is also recommend to remove outliers.
Based on empirical evaluation of the distribution an overestimation of transcripts count over the cells, we removed all cells with; 
Lung: >2000 transcripts; Spleen: >6000 transcripts; Muscle and Skin: >7000 transcripts; Joint: No outliers removed.

The output files are saved in data/sorted_DEGs, one file per organ.

```{r, echo=FALSE, tidy=TRUE, tidy.opts=list(width.cutoff=1)}

source('R/sc_data_quality_sorting.R')  
sc_data_quality_sorting(filenames[1], organ = 'Joint')  
sc_data_quality_sorting(filenames[2], organ = 'Muscle')  
sc_data_quality_sorting(filenames[3], organ = 'Lung')  
sc_data_quality_sorting(filenames[4], organ = 'Skin')  
sc_data_quality_sorting(filenames[5], organ = 'Spleen')

```



## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
