---
title: "CIA project"
author: "Sandra Lilja, Martin Smelink, Xinxiu Lee, Danuta Gawel, Oleg Sysoev"
date: "07/04/2022"
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


# scRNA-seq analysis for construction of MCDMs, MO-MCDMs, and Connective Pathway Analysis

## Define data structure and input files

```{r, echo FALSE, warning=FALSE, eval=FALSE}

if (dir.exists("data/IPA/curation")==FALSE){
  dir.create("data/IPA/curation", recursive = T)
}
if (dir.exists("data/sorted_DGEs")==FALSE){
  dir.create("data/sorted_DGEs")
}
if (dir.exists("data/NicheNet_analysis_curated")==FALSE){
  dir.create("data/NicheNet_analysis_curated")
}
if (dir.exists("data/clustering_and_celltype_identification")==FALSE){
  dir.create("data/clustering_and_celltype_identification")
}
if (dir.exists("data/scVI_normalized")==FALSE){
  dir.create("data/scVI_normalized")
}
```

The expected data input are five csv.gz matrices, one per organ, with cells over columns and genes over rows.

```{r}
### Read data from GSE and save it to data/ 
input_file = "data/UMI_expression_matrix.txt.gz"
```

## Quality assessment and sorting

To ensure good quality data for downstream analyses, it is recommended to remove poor quality cells and genes from the analyses.
We performed these analyses on the input_file, for all the organs combined. 
We define and keep the good quality cells as those having a minimum of 400 transcripts, 200 genes, and less than 20% mitochondrial genes. 
The good quality genes kept were defined as those being identified in at least 1% of the cells. 

Due to the risk of duplicates in the library resulting in two or more cells sharing a cell barcode, it is also recommend to remove outliers.
Based on empirical evaluation of the distribution an overestimation of transcripts count over the cells, we removed all cells with >6000 transcripts.

The output file is saved in data/sorted_DEGs.

```{r, eval=FALSE}
source('R/sc_data_quality_sorting.R')
sc_data_quality_sorting(input_file)
```


### Clustering and cell type identification by Leiden's algorithm

```{bash, eval=FALSE}
#module add anaconda3
#source activate dca
python3 Python/ConvertReferencesSandra.ipynb
#source deactivate
```

### Cell type identification using marker genes

As the reference used for cell type identification in above method does not contain all cell types, some clusters remained unidentified after this analysis. To verification the cell types which were identified, as well as to further identify the so far unidentified clusters, we ...

### Data normalization and Differential expression analysis

For each cell type in each organ, we calculate the DEGs between sick and healthy mice using scVI-tools (version 0.7.1)[^Gayoso]. The data was trained and normalized, including batch effect removal based on the individual samples sequenced (i.e. between organs and mice IDs). DEGs were then calculated between the two groups including the parameter mode='change'. Additionally, DEGs were calculated between each cell type/cluster (as defined by the [Leiden's algorithm](#clustering-and-cell-type-identification-by-leiden's-algorithm)) described above and all other cell types in the data. 

Note that the training will produce somewhat different output from each run, so in case the analyses need to be reproduced, the trained data is saved to 'outdir', '/full_posterior_400/model_params'. If this directory exists, the trained data will be loaded from there. As the training can be time consuming, this also saves computational time. Note also that the training results will depend on the previous set-ups, and if any of the parameters 'n_epochs', 'use_batches', or 'batch_id' are changed, the training need to be re-run.  

** Not provided yet, too big?** To reproduce our results, use the trained data provided in 'data/scVI_normalized/full_posterior_400/model_params'.

```{bash, eval=FALSE}
module add anaconda3
source activate cia # if needed, define user-specific python environment

# define in-/output directories and files
infile=data/sorted_DGEs/sorted_expression_matrix.csv
outdir=data/scVI_normalized/
clust_file=data/clustering_and_celltype_identification/cluster_ids.csv
outdir_DEG=data/DEG_analysis/sick_vs_healthy
mkdir -p $outdir_DEG
mkdir -p $outdir_DEG/change_mode

# Calculate DEGs for each cell type and organ, between sick and healthy mice
python3 Python/scVI_v0.7.1.py --infile $infile --outdir $outdir --n_epochs 400 --DEG_analysis --clust_file $clust_file --outdir_DEG $outdir_DEG --sort_column 1 3 --group_column 2 --use_batches --batch_id 0 2
wait

# Calculate DEGs for each cell type and organ between the cell type and all other cell types

```

```{bash}
module add anaconda3
source activate cia # if needed, define user-specific python environment
python3 Python/scVI_v0.7.1.py --help
```

The output from this code consists of one table of DEGs for each cell type and organ as defined by 'sort_column'  (see example output below), saved in 'outdir_DEGs'. These files contains values which describes the data; scale values from the training, log foldchange values, non-zero proportions, and raw normalized mean values, for group 1 and group 2 respectively. The group-ids can be interpreted based on the output file names, group1_vs_group2 (eg. sick_vs_healthy). We define the significant DEGs based on the column 'is_de_fdr_0.05'=True.

Additionally, the normalized and latent data is saved to 'outdir' for downstream analyses.  

```{r}
# Example output from scVI.v0.7.1.py
Joint_Mac <- read.csv('data/DEG_analysis/sick_vs_healthy/change_mode/DEGs_Sick_vs_Healthy_Macrophages_Joint.csv')
head(Joint_Mac)
```

We further subset these tables only to include significant DEGs. 

```{r}
source('R/DEG_sort_significant.R')
outdir_DEG <- 'data/DEG_analysis/sick_vs_healthy/change_mode'
DEG_files <- list.files(outdir_DEG, full.names = TRUE)
DEG_sort_significant.R(DEG_files)
```





### Identification of cellular interactions and MCDM/MO-MCDM construction

To identify cell-cell interactions, we used NicheNet[^Browaeys], a tool to model intercellular communications.
To define the genes of interest for identification of the intercellular interactions, we used the lists of DEGs for each cell type in each organ. 
The interactions were thus predicted, between each pair of cell types and each organ separately. 

As the NicheNet database consists of intercellular interactions in humans, the gene names were translated to their human orthologs. 
The translation file used for our studies is provided in /data/orthologous_translation_file.txt. 
To run this code on any other dataset, it is recommended to download the orthologs from Ensembl as described [here](https://www.ensembl.info/2009/01/21/how-to-get-all-the-orthologous-genes-between-two-species/).

The input files to run NicheNet_analysis.R, are;  

* the normalized expression data. Output from [scVI_v0.7.1.py](#data-normalization-and-differential-expression-analysis)
* the cell type identification file. Output from [cell type analysis script](#clustering-and-cell-type-identification)
* a translation file of human orthologs
* a matrix listing all DEGs for each cell type and organ combination, with celltype_organ over columns, as output from the [differential expression analysis](#data-normalization-and-differential-expression-analysis), DEG_sort_significant.R. 

The code output one tab separated txt file per interacting cell type pair, within and between organs, 
containing "test_ligand" (i.e upstream regulator of interaction), "auroc", "aupr", "pearson" (Pearson Correlation Coefficient, PCC), "target" (genes), and "target_weight" over the columns and interactions over rows. 

Additionally, it creates one file containing all predicted interactions between each pair of cell types, all_ligand_activity.txt, 
which apart from above mentioned columns contains "Sender" and "Target" cell type and organ (named: celltype_organ)

Note: that the interactions between organs in these files are not yet curated only to include URs secreated in blood

To curate the inter-organ interactions, only to include interactions through URs secreted in blood, 
and to sort out the strongest interactions (PCC > 0) considered for further analyses
we ran NicheNet_network_curation.R. As input to this script is the all_ligand_activity.txt output from NicheNet_analysis.R, 
and a curation file (eg., data/IPA/curation/curation_file.txt) containing information about the cellular location of each UR. 

The curation file can be retrieved from IPA by following [IPA - Generate curation file](#generate-curation-file)

It is important that the curation file contains two columns named "Symbol" (containing the gene names of the URs) and "Location" (containing information about cellular location). 
The URs which will be kept for inter-organ interactions by the code are those with "Location" == "Extracellular Space". 

The output from this script consists of one file containing all inter- and intra-organ interactions, all_pos_curated_ligand_activity.txt (Fig 1)


### Ranking of URs based on their downstream effect

/codes/rank_by_targets_and_heatmap.R

### Connective Pathway Analysis

## Meta analysis of 11 IMIDs

### Differential expression analysis

### ...

## Ingenuity pathway analysis

### Generate curation file

### Pathway analysis

### Prediction of upstream regulators (URs)

# References
[^Gayoso]: Gayoso, Adam, et al. A Python library for probabilistic analysis of single-cell omics data. *Nature Biotechnology* 40.2: 163-166, doi:10.1038/s41587-021-01206-w (2022).
[^Browaeys]: Browaeys, R., Saelens, W. & Saeys, Y. NicheNet: modeling intercellular communication by linking ligands to target genes. *Nature Methods* 17, 159-162, doi:10.1038/s41592-019-0667-5 (2020).
