% This LaTeX was auto-generated from MATLAB code.
% To make changes, update the MATLAB code and export to LaTeX again.

\documentclass{article}

\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{lmodern}
\usepackage{graphicx}
\usepackage{color}
\usepackage{hyperref}
\usepackage{amsmath}
\usepackage{amsfonts}
\usepackage{epstopdf}
\usepackage[table]{xcolor}
\usepackage{matlab}

\sloppy
\epstopdfsetup{outdir=./}
\graphicspath{ {./UR_enrichment_CIA_images/} }

\matlabhastoc

\begin{document}

\label{T_3BBCC1A2}
\matlabtitle{UR enrichment analysis of CIA}

\matlabtableofcontents{Table of Contents}

\vspace{1em}
\begin{matlabcode}
clear all
close all
clc
warning('off', 'all')
\end{matlabcode}


\label{H_ABA800F9}
\matlabheading{Input requirements}

\label{H_29D9EA9B}
\matlabheadingthree{For this analysis we will need:}

\begin{itemize}
\setlength{\itemsep}{-1ex}
   \item{\begin{flushleft} DEGs computed by (TODO link!) \end{flushleft}}
   \item{\begin{flushleft} Orthologous translation \end{flushleft}}
   \item{\begin{flushleft} Results from the connective pathway analysis (TODO link!) \end{flushleft}}
   \item{\begin{flushleft} Predicted URs and their downstreamers from Niche Net (TODO link!) \end{flushleft}}
\end{itemize}

\label{H_229A39C3}
\matlabheadingthree{We can find the files at following locations:}

\begin{par}
\hfill \break
\end{par}

\begin{matlabcode}
% DEGs
path_DEGs = '../data/CIA_mouse_DEGs/';
% translation
path_translation = '../data/Connective_pathway_analysis/orthologous_translation_NicheNet_analysis.txt';
% connective pathway analysis
path_cpa = '../data/Connective_pathway_analysis/TreeStructure_nodes2_scRNAseq.txt';
% URs 
path_URs = '../data/Connective_pathway_analysis/all_outgoing_ligand_activity.txt';
\end{matlabcode}

\label{H_77554A53}
\matlabheadingthree{We will produce}

\begin{par}
\begin{flushleft}
A table with combined, FDR corrected pvalues and the count of significant p values across all cell types for each program/subprogram and UR 
\end{flushleft}
\end{par}

\label{H_CD1D8F27}
\matlabheadingthree{We will save the file here:}

\begin{par}
\hfill \break
\end{par}

\begin{matlabcode}
path_output = '../data/UR_analysis/UR_predictions_CIA_disease_Pvals.xlsx';
\end{matlabcode}


\vspace{1em}

\label{H_8EEC0AFA}
\matlabheading{Prepare the data}


\vspace{1em}
\label{H_C9AE6352}
\matlabheadingthree{Load DEGs and translate the gene names to human orthologous.}

\begin{par}
\hfill \break
\end{par}

\begin{matlabcode}

Fn = struct2cell(dir(path_DEGs));
Fn(2:end,:)=[];
Fn(~contains(Fn,'.csv'))=[];
Fn = table(Fn','variablenames',{'fn'});
Fn.Cell = cellfun(@(x) x{5}, regexp(Fn.fn,'\_','split'),'UniformOutput',0);
Fn.Organ = cellfun(@(x) x{6}(1:end-4), regexp(Fn.fn,'\_','split'),'UniformOutput',0);
Fn(~ismember(Fn.Organ,{'Joint','Muscle'}),:)=[];

translation = readtable(path_translation);
for z = 1 : length(Fn.fn)
   hv =readtable(sprintf('%s%s',path_DEGs,Fn.fn{z}));
   hv(~strcmp(hv.is_de_fdr_0_05,'True'),:)=[];
   xx= unique(translation(ismember(translation.Mouse_gene_name,hv.Var1),:));
   Fn.DEG{z} = xx.Gene_name;
   clear hv xx
end

% Restructure the data
Fn.cell_dataset = cellfun(@(x) x{5},regexp(Fn.fn,'\_','split'),'UniformOutput',0);
Fn.Properties.VariableNames(strcmp(Fn.Properties.VariableNames,'Organ'))={'Organ_condition'};
Fn.group = repmat({'Joint'},length(Fn.fn),1);
Fn.group(strcmp(Fn.Organ_condition,'Muscle'))={'Muscle'};
Fn.key = cellfun(@(x,y) sprintf('%s***%s',x,y),Fn.cell_dataset,Fn.group,'UniformOutput',0);

Fn = Fn(:,{'key','cell_dataset','group','DEG'});

head(Fn, 5)
\end{matlabcode}
\begin{matlabtableoutput}
{
\begin{tabular} {|c|c|c|c|c|}\hline
\mlcell{ } & \mlcell{key} & \mlcell{cell\_dataset} & \mlcell{group} & \mlcell{DEG} \\ \hline
\mlcell{1} & \mlcell{'B-cells***Muscle'} & \mlcell{'B-cells'} & \mlcell{'Muscle'} & \mlcell{2919x1 cell} \\ \hline
\mlcell{2} & \mlcell{'Dendritic-cells***Muscle'} & \mlcell{'Dendritic-cells'} & \mlcell{'Muscle'} & \mlcell{2516x1 cell} \\ \hline
\mlcell{3} & \mlcell{'Endothelial-cells***Joint'} & \mlcell{'Endothelial-cells'} & \mlcell{'Joint'} & \mlcell{1737x1 cell} \\ \hline
\mlcell{4} & \mlcell{'Endothelial-cells***Muscle'} & \mlcell{'Endothelial-cells'} & \mlcell{'Muscle'} & \mlcell{2460x1 cell} \\ \hline
\mlcell{5} & \mlcell{'Erythrocytes***Muscle'} & \mlcell{'Erythrocytes'} & \mlcell{'Muscle'} & \mlcell{3211x1 cell} \\ 
\hline
\end{tabular}
}
\end{matlabtableoutput}

\label{H_CCE50919}
\matlabheadingthree{Load CIA\_Ps and CIA\_SPs from connective pathway analysis}

\begin{par}
\hfill \break
\end{par}

\begin{matlabcode}


SPhv = readtable(path_cpa);
SPhv.subprograms = cellfun(@(x) sprintf('1.%s',x), strtrim(cellstr(num2str(SPhv.subclusters))),'UniformOutput',0);
SP = table(unique(SPhv.subprograms),'variablenames',{'SP'});
for z = 1 : length(SP.SP)
    SP.AllMolecules{z} = unique(regexp(strjoin(SPhv.AllMolecules(strcmp(SPhv.subprograms,SP.SP{z})),','),'\,','split'));
end

SP.SP(length(SP.SP)+1)={'P1'};
SP.AllMolecules{end} = unique(regexp(strjoin(SPhv.AllMolecules(SPhv.clustersGlobal==1),','),'\,','split'));
SP.SP(length(SP.SP)+1)={'P2'};
SP.AllMolecules{end} = unique(regexp(strjoin(SPhv.AllMolecules(SPhv.clustersGlobal==2),','),'\,','split'));

head(SP, 5)
\end{matlabcode}
\begin{matlabtableoutput}
{
\begin{tabular} {|c|c|c|}\hline
\mlcell{ } & \mlcell{SP} & \mlcell{AllMolecules} \\ \hline
\mlcell{1} & \mlcell{'1.1'} & \mlcell{1x393 cell} \\ \hline
\mlcell{2} & \mlcell{'1.10'} & \mlcell{1x108 cell} \\ \hline
\mlcell{3} & \mlcell{'1.2'} & \mlcell{1x716 cell} \\ \hline
\mlcell{4} & \mlcell{'1.3'} & \mlcell{1x247 cell} \\ \hline
\mlcell{5} & \mlcell{'1.4'} & \mlcell{1x351 cell} \\ 
\hline
\end{tabular}
}
\end{matlabtableoutput}


\vspace{1em}

\label{H_BBBEC1DE}
\matlabheadingthree{load URs from Niche net}

\begin{par}
\hfill \break
\end{par}

\begin{matlabcode}


UR = readtable(path_URs);
UR(UR.pearson<0,:)=[];
UR.senderOrgan = cellfun(@(x) x{2}, regexp(UR.Sender,'\_','split'),'UniformOutput',0);
UR.targetOrgan = cellfun(@(x) x{2}, regexp(UR.Target,'\_','split'),'UniformOutput',0);
UR(~strcmp(UR.senderOrgan,UR.targetOrgan),:)=[];
UR(~ismember(UR.senderOrgan,{'Muscle','Joint'}),:)=[];
UR.targetcell = cellfun(@(x) x{1}, regexp(UR.Target,'\_','split'),'UniformOutput',0);
UR = unique(UR(:,{'test_ligand','targetcell','targetOrgan','target'}));
UR.Properties.VariableNames = {'UR','cell_dataset','condition','DS'};
UR.DS = regexp(UR.DS,'\/','split');
UR.group(strcmp(UR.condition,'Muscle')) = {'Muscle'};
UR.group(strcmp(UR.condition,'Joint')) = {'Joint'};
UR.key = cellfun(@(x,y) sprintf('%s***%s',x,y),UR.cell_dataset,UR.group,'UniformOutput',0);
UR = UR(:,{'UR','key','cell_dataset','group','DS'});
head(UR, 5)
\end{matlabcode}
\begin{matlabtableoutput}
{
\begin{tabular} {|c|c|c|c|c|c|}\hline
\mlcell{ } & \mlcell{UR} & \mlcell{key} & \mlcell{cell\_dataset} & \mlcell{group} & \mlcell{DS} \\ \hline
\mlcell{1} & \mlcell{'APOE'} & \mlcell{'B-cells***Muscle'} & \mlcell{'B-cells'} & \mlcell{'Muscle'} & \mlcell{1x64 cell} \\ \hline
\mlcell{2} & \mlcell{'APOE'} & \mlcell{'Dendritic-cells***Muscle'} & \mlcell{'Dendritic-cells'} & \mlcell{'Muscle'} & \mlcell{1x50 cell} \\ \hline
\mlcell{3} & \mlcell{'APOE'} & \mlcell{'Erythrocytes***Muscle'} & \mlcell{'Erythrocytes'} & \mlcell{'Muscle'} & \mlcell{1x69 cell} \\ \hline
\mlcell{4} & \mlcell{'APOE'} & \mlcell{'Granulocytes***Muscle'} & \mlcell{'Granulocytes'} & \mlcell{'Muscle'} & \mlcell{1x24 cell} \\ \hline
\mlcell{5} & \mlcell{'APOE'} & \mlcell{'Macrophages***Muscle'} & \mlcell{'Macrophages'} & \mlcell{'Muscle'} & \mlcell{1x18 cell} \\ 
\hline
\end{tabular}
}
\end{matlabtableoutput}


\vspace{1em}

\label{H_C543E7C0}
\matlabheading{Main analysis }

\label{H_DB504A76}
\matlabheadingthree{Fisher test enrichment of UR DS in SPs:}

\begin{par}
\hfill \break
\end{par}

\begin{matlabcode}

FPval = cell(length(SP.SP),1);
FOR = cell(length(SP.SP),1);


uUR = unique(UR.UR);
for p = 1 : length(SP.SP)
    for cd = 1 : length(Fn.key)               
            spgenes = SP.AllMolecules{p}(ismember(SP.AllMolecules{p},Fn.DEG{cd}));
            URhv = UR(strcmp(UR.key,Fn.key{cd}),:); %!!!!! URs from more datasets - solved .key now correspond to datasets

            for ur = 1 : length(uUR)
                if sum(strcmp(URhv.UR,uUR{ur}))
                    if length(unique(Fn.DEG{cd})) ~= length(Fn.DEG{cd})
                        error = p;
                    end

                    urds = URhv.DS{strcmp(URhv.UR,uUR{ur})}; 

                    a = sum(ismember(spgenes,urds));
                    b = sum(~ismember(spgenes,urds));
                    c = sum(~ismember(urds,spgenes));
                    d = sum(~ismember(Fn.DEG{cd},[spgenes,urds]));

                    [~, FPval{p}(cd,ur)] = fishertest([a b; c d],'tail','right');
                    FOR{p}(cd,ur) = (a*d)/(b*c);
                else
                    FPval{p}(cd,ur) = 1;
                    FOR{p}(cd,ur) = 0;
                end

                clear urds a b c d
            end
            clear spgenes
    end
end

% Summarize results
FishertestR.cellsdes = SP;
FishertestR.rowdes = Fn;
FishertestR.coldes_URs = uUR;
FishertestR.moreURinfo = UR;
FishertestR.Pval = FPval;
FishertestR.OR = FOR;
FishertestR
\end{matlabcode}
\begin{matlaboutput}
FishertestR = 
      cellsdes: [12x2 table]
        rowdes: [18x4 table]
    coldes_URs: {31x1 cell}
    moreURinfo: [123x5 table]
          Pval: {12x1 cell}
            OR: {12x1 cell}

\end{matlaboutput}


\vspace{1em}

\label{H_2E8C639C}
\matlabheadingthree{combine pvalues over joint and muscle:}

\begin{par}
\hfill \break
\end{par}

\begin{matlabcode}


% For each program and UR, use Fisher's method to combine pvals over all cell types 
for p = 1 : length(FishertestR.Pval)
    for ur = 1 : size(FishertestR.Pval{p},2)
        joint_idx = strcmp(FishertestR.rowdes.group,'Joint');
        chi = -2*sum(log(FishertestR.Pval{p}(joint_idx,ur)));
        CombinedPval_joint(p,ur) = my_chi2cdf(chi,2*sum(joint_idx));
        
        muscle_idx = strcmp(FishertestR.rowdes.group,'Muscle');
        chi = -2*sum(log(FishertestR.Pval{p}(muscle_idx,ur)));
        CombinedPval_muscle(p,ur) = my_chi2cdf(chi,2*sum(muscle_idx));
        
        chi = -2*sum(log(FishertestR.Pval{p}(:,ur)));
        CombinedPval_all(p,ur) = my_chi2cdf(chi,2*32);
    end
end


% Create a summary file where we store important results
CombinedOverJointandMuscle.cols_URs = FishertestR.coldes_URs;
CombinedOverJointandMuscle.rows_SPs = FishertestR.cellsdes.SP;
CombinedOverJointandMuscle.CombP_joint = CombinedPval_joint;
CombinedOverJointandMuscle.CombP_muscle = CombinedPval_muscle;
CombinedOverJointandMuscle.CombP_all = CombinedPval_all;
CombinedOverJointandMuscle
\end{matlabcode}
\begin{matlaboutput}
CombinedOverJointandMuscle = 
        cols_URs: {31x1 cell}
        rows_SPs: {12x1 cell}
     CombP_joint: [12x31 double]
    CombP_muscle: [12x31 double]
       CombP_all: [12x31 double]

\end{matlaboutput}


\vspace{1em}

\label{H_BC696F3B}
\matlabheadingthree{FDR correction over Programs}

\begin{par}
\hfill \break
\end{par}

\begin{matlabcode}


% For each program and UR, do FDR correction 
for i = 1:length(CombinedOverJointandMuscle.CombP_joint(:,1))
    FDR_joint(i,:) = mafdr(CombinedOverJointandMuscle.CombP_joint(i,:), 'BHFDR',true);
    FDR_muscle(i,:) = mafdr(CombinedOverJointandMuscle.CombP_muscle(i,:), 'BHFDR',true);
    FDR_all(i,:) = mafdr(CombinedOverJointandMuscle.CombP_all(i,:), 'BHFDR',true);

end

CombinedOverJointandMuscle.FDR_Joint = FDR_joint;
CombinedOverJointandMuscle.FDR_Muscle = FDR_muscle;
CombinedOverJointandMuscle.FDR_All = FDR_all;

CombinedOverJointandMuscle
\end{matlabcode}
\begin{matlaboutput}
CombinedOverJointandMuscle = 
        cols_URs: {31x1 cell}
        rows_SPs: {12x1 cell}
     CombP_joint: [12x31 double]
    CombP_muscle: [12x31 double]
       CombP_all: [12x31 double]
       FDR_Joint: [12x31 double]
      FDR_Muscle: [12x31 double]
         FDR_All: [12x31 double]

\end{matlaboutput}


\vspace{1em}

\vspace{1em}

\label{H_0BDB29FB}
\matlabheadingthree{Count significant pvalues over joints and muscles:}

\begin{par}
\hfill \break
\end{par}

\begin{matlabcode}


% For each program and UR, count how in how many cell types was UR significant 
for p = 1 : length(FishertestR.Pval)
    for ur = 1 : size(FishertestR.Pval{p},2)
        joint_idx = strcmp(FishertestR.rowdes.group,'Joint');
        pvals_corrected_joint = mafdr(FishertestR.Pval{p}(joint_idx,ur), 'BHFDR',true);
        count_joint(p,ur) = sum(pvals_corrected_joint<0.05);
        %datasets_joint(p,ur,:) = pvals_corrected_joint<0.05;
        
        muscle_idx = strcmp(FishertestR.rowdes.group,'Muscle');
        pvals_corrected_muscle = mafdr(FishertestR.Pval{p}(muscle_idx,ur), 'BHFDR',true);
        count_muscle(p,ur) = sum(pvals_corrected_muscle<0.05);
        %datasets_muscle(p,ur,:) = pvals_corrected_muscle<0.05;

    end
end

CombinedOverJointandMuscle.count_joint = count_joint;
CombinedOverJointandMuscle.count_muscle = count_muscle;

CombinedOverJointandMuscle
\end{matlabcode}
\begin{matlaboutput}
CombinedOverJointandMuscle = 
        cols_URs: {31x1 cell}
        rows_SPs: {12x1 cell}
     CombP_joint: [12x31 double]
    CombP_muscle: [12x31 double]
       CombP_all: [12x31 double]
       FDR_Joint: [12x31 double]
      FDR_Muscle: [12x31 double]
         FDR_All: [12x31 double]
     count_joint: [12x31 double]
    count_muscle: [12x31 double]

\end{matlaboutput}


\vspace{1em}

\label{H_C7F41035}
\matlabheading{Final output}

\begin{matlabcode}

CombinedPval  = table(repmat(CombinedOverJointandMuscle.rows_SPs,size(CombinedOverJointandMuscle.CombP_joint,2),1),...
                    repelem(CombinedOverJointandMuscle.cols_URs,size(CombinedOverJointandMuscle.CombP_joint,1)),...
                    reshape(CombinedOverJointandMuscle.CombP_joint,size(CombinedOverJointandMuscle.CombP_joint,1)*size(CombinedOverJointandMuscle.CombP_joint,2),1),...
                    reshape(CombinedOverJointandMuscle.CombP_muscle,size(CombinedOverJointandMuscle.CombP_muscle,1)*size(CombinedOverJointandMuscle.CombP_muscle,2),1),...
                    reshape(CombinedOverJointandMuscle.CombP_all,size(CombinedOverJointandMuscle.CombP_all,1)*size(CombinedOverJointandMuscle.CombP_all,2),1),...
                    reshape(CombinedOverJointandMuscle.FDR_Joint,size(CombinedOverJointandMuscle.FDR_Joint,1)*size(CombinedOverJointandMuscle.FDR_Joint,2),1),...
                    reshape(CombinedOverJointandMuscle.FDR_Muscle,size(CombinedOverJointandMuscle.FDR_Muscle,1)*size(CombinedOverJointandMuscle.FDR_Muscle,2),1),...
                    reshape(CombinedOverJointandMuscle.FDR_All,size(CombinedOverJointandMuscle.FDR_All,1)*size(CombinedOverJointandMuscle.FDR_All,2),1),...
                    reshape(CombinedOverJointandMuscle.count_joint,size(CombinedOverJointandMuscle.count_joint,1)*size(CombinedOverJointandMuscle.count_joint,2),1),...
                    reshape(CombinedOverJointandMuscle.count_muscle,size(CombinedOverJointandMuscle.count_muscle,1)*size(CombinedOverJointandMuscle.count_muscle,2),1),...
                    'variablenames',{'SP','UR','CombinedP_Joint','CombinedP_Muscle', 'CombinedP_All', 'FDR_Joint', 'FDR_Muscle', 'FDR_All', 'count_Joint', 'count_Muscle'});
head(CombinedPval, 5)
\end{matlabcode}
\begin{matlabtableoutput}
{
\begin{tabular} {|c|c|c|c|c|c|c|c|c|c|c|}\hline
\mlcell{ } & \mlcell{SP} & \mlcell{UR} & \mlcell{CombinedP\_Joint} & \mlcell{CombinedP\_Muscle} & \mlcell{CombinedP\_All} & \mlcell{FDR\_Joint} & \mlcell{FDR\_Muscle} & \mlcell{FDR\_All} & \mlcell{count\_Joint} & \mlcell{count\_Muscle} \\ \hline
\mlcell{1} & \mlcell{'1.1'} & \mlcell{'APOE'} & \mlcell{1.0000e+00} & \mlcell{4.5845e-20} & \mlcell{9.1843e-09} & \mlcell{1.0000e+00} & \mlcell{1.4212e-18} & \mlcell{2.8471e-07} & \mlcell{0} & \mlcell{7} \\ \hline
\mlcell{2} & \mlcell{'1.10'} & \mlcell{'APOE'} & \mlcell{1.0000e+00} & \mlcell{9.9784e-01} & \mlcell{1.0000e+00} & \mlcell{1.0000e+00} & \mlcell{1.0000e+00} & \mlcell{1.0000e+00} & \mlcell{0} & \mlcell{0} \\ \hline
\mlcell{3} & \mlcell{'1.2'} & \mlcell{'APOE'} & \mlcell{1.0000e+00} & \mlcell{1.9419e-22} & \mlcell{1.9182e-10} & \mlcell{1.0000e+00} & \mlcell{1.5050e-21} & \mlcell{1.4866e-09} & \mlcell{0} & \mlcell{7} \\ \hline
\mlcell{4} & \mlcell{'1.3'} & \mlcell{'APOE'} & \mlcell{1.0000e+00} & \mlcell{1.0711e-26} & \mlcell{1.3463e-13} & \mlcell{1.0000e+00} & \mlcell{1.1068e-25} & \mlcell{1.0434e-12} & \mlcell{0} & \mlcell{7} \\ \hline
\mlcell{5} & \mlcell{'1.4'} & \mlcell{'APOE'} & \mlcell{1.0000e+00} & \mlcell{2.6564e-24} & \mlcell{8.3587e-12} & \mlcell{1.0000e+00} & \mlcell{2.7449e-23} & \mlcell{6.4780e-11} & \mlcell{0} & \mlcell{7} \\ 
\hline
\end{tabular}
}
\end{matlabtableoutput}


\vspace{1em}

\label{H_2899577C}
\matlabheadingthree{Save the data}

\begin{par}
\hfill \break
\end{par}

\begin{matlabcode}
%writetable(CombinedPval,path_output)

\end{matlabcode}

\end{document}
