{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import gzip\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "variant_disease = pd.read_csv(\"../SAR/all_variant_disease_associations.tsv.gz\", compression='gzip', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find GWAS genes in the database\n",
    "RA = list()\n",
    "for i in range(len(variant_disease['diseaseName'])):\n",
    "    if 'rheumatoid' in variant_disease['diseaseName'][i]:\n",
    "        RA.append(i)\n",
    "    if 'Rheumatoid' in variant_disease['diseaseName'][i]:\n",
    "        RA.append(i)\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Early Rheumatoid Arthritis', 'Juvenile rheumatoid arthritis',\n",
       "       'Polyarticular Juvenile Idiopathic Arthritis, Rheumatoid Factor Negative',\n",
       "       'Progression of rheumatoid arthritis',\n",
       "       'Progressive pseudorheumatoid dysplasia', 'Rheumatoid Arthritis',\n",
       "       'Rheumatoid Arthritis, Systemic Juvenile',\n",
       "       'Rheumatoid Factor Measurement', 'Rheumatoid Nodule',\n",
       "       'Seropositive rheumatoid arthritis'], dtype=object)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.unique(variant_disease['diseaseName'][RA])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Let's say that we don't want Pediatric Crohn's disease, we can get rid of it for example the following way\n",
    "RA = list()\n",
    "for i in range(len(variant_disease['diseaseName'])):\n",
    "    if 'Rheumatoid Arthritis' in variant_disease['diseaseName'][i]\\\n",
    "    and 'Juvenile' not in variant_disease['diseaseName'][i]\\\n",
    "    and 'Early' not in variant_disease['diseaseName'][i]:\n",
    "        RA.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Rheumatoid Arthritis'], dtype=object)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Now we are satisfied with the diseases\n",
    "np.unique(variant_disease['diseaseName'][RA])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snpId</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>DSI</th>\n",
       "      <th>DPI</th>\n",
       "      <th>diseaseId</th>\n",
       "      <th>diseaseName</th>\n",
       "      <th>diseaseType</th>\n",
       "      <th>diseaseClass</th>\n",
       "      <th>diseaseSemanticType</th>\n",
       "      <th>score</th>\n",
       "      <th>EI</th>\n",
       "      <th>YearInitial</th>\n",
       "      <th>YearFinal</th>\n",
       "      <th>NofPmids</th>\n",
       "      <th>source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>608</th>\n",
       "      <td>rs1003878</td>\n",
       "      <td>6</td>\n",
       "      <td>32332045</td>\n",
       "      <td>0.882</td>\n",
       "      <td>0.20</td>\n",
       "      <td>C0003873</td>\n",
       "      <td>Rheumatoid Arthritis</td>\n",
       "      <td>disease</td>\n",
       "      <td>C17;C05;C20</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2007.0</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>3</td>\n",
       "      <td>GWASDB;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>611</th>\n",
       "      <td>rs1003879</td>\n",
       "      <td>6</td>\n",
       "      <td>32331815</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.12</td>\n",
       "      <td>C0003873</td>\n",
       "      <td>Rheumatoid Arthritis</td>\n",
       "      <td>disease</td>\n",
       "      <td>C17;C05;C20</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>2011.0</td>\n",
       "      <td>1</td>\n",
       "      <td>GWASDB;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>758</th>\n",
       "      <td>rs1004819</td>\n",
       "      <td>1</td>\n",
       "      <td>67204530</td>\n",
       "      <td>0.776</td>\n",
       "      <td>0.36</td>\n",
       "      <td>C0003873</td>\n",
       "      <td>Rheumatoid Arthritis</td>\n",
       "      <td>disease</td>\n",
       "      <td>C17;C05;C20</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>2013.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>809</th>\n",
       "      <td>rs1005133</td>\n",
       "      <td>22</td>\n",
       "      <td>19750832</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.12</td>\n",
       "      <td>C0003873</td>\n",
       "      <td>Rheumatoid Arthritis</td>\n",
       "      <td>disease</td>\n",
       "      <td>C17;C05;C20</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>1</td>\n",
       "      <td>GWASCAT;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>916</th>\n",
       "      <td>rs1005753</td>\n",
       "      <td>1</td>\n",
       "      <td>17118274</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.12</td>\n",
       "      <td>C0003873</td>\n",
       "      <td>Rheumatoid Arthritis</td>\n",
       "      <td>disease</td>\n",
       "      <td>C17;C05;C20</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.01</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>1</td>\n",
       "      <td>BEFREE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>367688</th>\n",
       "      <td>rs9880772</td>\n",
       "      <td>3</td>\n",
       "      <td>27736288</td>\n",
       "      <td>0.827</td>\n",
       "      <td>0.24</td>\n",
       "      <td>C0003873</td>\n",
       "      <td>Rheumatoid Arthritis</td>\n",
       "      <td>disease</td>\n",
       "      <td>C17;C05;C20</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>2015.0</td>\n",
       "      <td>1</td>\n",
       "      <td>GWASCAT;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368137</th>\n",
       "      <td>rs9909240</td>\n",
       "      <td>17</td>\n",
       "      <td>9236774</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.12</td>\n",
       "      <td>C0003873</td>\n",
       "      <td>Rheumatoid Arthritis</td>\n",
       "      <td>disease</td>\n",
       "      <td>C17;C05;C20</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>1</td>\n",
       "      <td>GWASDB;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>368286</th>\n",
       "      <td>rs991760</td>\n",
       "      <td>6</td>\n",
       "      <td>32855790</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.12</td>\n",
       "      <td>C0003873</td>\n",
       "      <td>Rheumatoid Arthritis</td>\n",
       "      <td>disease</td>\n",
       "      <td>C17;C05;C20</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.70</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>2009.0</td>\n",
       "      <td>1</td>\n",
       "      <td>GWASDB;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369265</th>\n",
       "      <td>rs9979383</td>\n",
       "      <td>21</td>\n",
       "      <td>35343463</td>\n",
       "      <td>0.925</td>\n",
       "      <td>0.20</td>\n",
       "      <td>C0003873</td>\n",
       "      <td>Rheumatoid Arthritis</td>\n",
       "      <td>disease</td>\n",
       "      <td>C17;C05;C20</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.80</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2012.0</td>\n",
       "      <td>2016.0</td>\n",
       "      <td>3</td>\n",
       "      <td>GWASCAT;GWASDB;INFERRED</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>369366</th>\n",
       "      <td>rs998731</td>\n",
       "      <td>8</td>\n",
       "      <td>80183160</td>\n",
       "      <td>1.000</td>\n",
       "      <td>0.12</td>\n",
       "      <td>C0003873</td>\n",
       "      <td>Rheumatoid Arthritis</td>\n",
       "      <td>disease</td>\n",
       "      <td>C17;C05;C20</td>\n",
       "      <td>Disease or Syndrome</td>\n",
       "      <td>0.80</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>2019.0</td>\n",
       "      <td>2</td>\n",
       "      <td>GWASCAT;GWASDB;INFERRED</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2387 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            snpId chromosome  position    DSI   DPI diseaseId  \\\n",
       "608     rs1003878          6  32332045  0.882  0.20  C0003873   \n",
       "611     rs1003879          6  32331815  1.000  0.12  C0003873   \n",
       "758     rs1004819          1  67204530  0.776  0.36  C0003873   \n",
       "809     rs1005133         22  19750832  1.000  0.12  C0003873   \n",
       "916     rs1005753          1  17118274  1.000  0.12  C0003873   \n",
       "...           ...        ...       ...    ...   ...       ...   \n",
       "367688  rs9880772          3  27736288  0.827  0.24  C0003873   \n",
       "368137  rs9909240         17   9236774  1.000  0.12  C0003873   \n",
       "368286   rs991760          6  32855790  1.000  0.12  C0003873   \n",
       "369265  rs9979383         21  35343463  0.925  0.20  C0003873   \n",
       "369366   rs998731          8  80183160  1.000  0.12  C0003873   \n",
       "\n",
       "                 diseaseName diseaseType diseaseClass  diseaseSemanticType  \\\n",
       "608     Rheumatoid Arthritis     disease  C17;C05;C20  Disease or Syndrome   \n",
       "611     Rheumatoid Arthritis     disease  C17;C05;C20  Disease or Syndrome   \n",
       "758     Rheumatoid Arthritis     disease  C17;C05;C20  Disease or Syndrome   \n",
       "809     Rheumatoid Arthritis     disease  C17;C05;C20  Disease or Syndrome   \n",
       "916     Rheumatoid Arthritis     disease  C17;C05;C20  Disease or Syndrome   \n",
       "...                      ...         ...          ...                  ...   \n",
       "367688  Rheumatoid Arthritis     disease  C17;C05;C20  Disease or Syndrome   \n",
       "368137  Rheumatoid Arthritis     disease  C17;C05;C20  Disease or Syndrome   \n",
       "368286  Rheumatoid Arthritis     disease  C17;C05;C20  Disease or Syndrome   \n",
       "369265  Rheumatoid Arthritis     disease  C17;C05;C20  Disease or Syndrome   \n",
       "369366  Rheumatoid Arthritis     disease  C17;C05;C20  Disease or Syndrome   \n",
       "\n",
       "        score   EI  YearInitial  YearFinal  NofPmids                   source  \n",
       "608      0.70  1.0       2007.0     2011.0         3          GWASDB;INFERRED  \n",
       "611      0.70  1.0       2011.0     2011.0         1          GWASDB;INFERRED  \n",
       "758      0.01  1.0       2013.0     2013.0         1                   BEFREE  \n",
       "809      0.70  1.0       2019.0     2019.0         1         GWASCAT;INFERRED  \n",
       "916      0.01  1.0       2012.0     2012.0         1                   BEFREE  \n",
       "...       ...  ...          ...        ...       ...                      ...  \n",
       "367688   0.70  1.0       2015.0     2015.0         1         GWASCAT;INFERRED  \n",
       "368137   0.70  1.0       2009.0     2009.0         1          GWASDB;INFERRED  \n",
       "368286   0.70  1.0       2009.0     2009.0         1          GWASDB;INFERRED  \n",
       "369265   0.80  1.0       2012.0     2016.0         3  GWASCAT;GWASDB;INFERRED  \n",
       "369366   0.80  1.0       2014.0     2019.0         2  GWASCAT;GWASDB;INFERRED  \n",
       "\n",
       "[2387 rows x 16 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#These are information about the diseases from the database\n",
    "variant_disease.loc[RA]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>snpId</th>\n",
       "      <th>geneId</th>\n",
       "      <th>geneSymbol</th>\n",
       "      <th>sourceId</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>rs1801280</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>VEP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>rs1799930</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>VEP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>rs1799931</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>DBSNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>rs1799931</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>VEP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>rs1799930</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>DBSNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361465</th>\n",
       "      <td>rs629326</td>\n",
       "      <td>112267968</td>\n",
       "      <td>LOC112267968</td>\n",
       "      <td>DBSNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361466</th>\n",
       "      <td>rs212389</td>\n",
       "      <td>112267968</td>\n",
       "      <td>LOC112267968</td>\n",
       "      <td>DBSNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361469</th>\n",
       "      <td>rs2451258</td>\n",
       "      <td>112267968</td>\n",
       "      <td>LOC112267968</td>\n",
       "      <td>DBSNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361916</th>\n",
       "      <td>rs2395379</td>\n",
       "      <td>113523636</td>\n",
       "      <td>SMIM40</td>\n",
       "      <td>DBSNP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361917</th>\n",
       "      <td>rs2073524</td>\n",
       "      <td>113523636</td>\n",
       "      <td>SMIM40</td>\n",
       "      <td>VEP</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4014 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            snpId     geneId    geneSymbol sourceId\n",
       "29      rs1801280         10          NAT2      VEP\n",
       "30      rs1799930         10          NAT2      VEP\n",
       "32      rs1799931         10          NAT2    DBSNP\n",
       "33      rs1799931         10          NAT2      VEP\n",
       "34      rs1799930         10          NAT2    DBSNP\n",
       "...           ...        ...           ...      ...\n",
       "361465   rs629326  112267968  LOC112267968    DBSNP\n",
       "361466   rs212389  112267968  LOC112267968    DBSNP\n",
       "361469  rs2451258  112267968  LOC112267968    DBSNP\n",
       "361916  rs2395379  113523636        SMIM40    DBSNP\n",
       "361917  rs2073524  113523636        SMIM40      VEP\n",
       "\n",
       "[4014 rows x 4 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Map the database to the geneSymbol names\n",
    "snpId = variant_disease.loc[RA]['snpId']\n",
    "variant_gene = pd.read_csv(\"../SAR/variant_to_gene_mappings.tsv.gz\", compression='gzip', sep = '\\t')\n",
    "chosen = list()\n",
    "for i in range(len(variant_gene['snpId'])):\n",
    "    if variant_gene['snpId'][i] in np.array(snpId):\n",
    "        chosen.append(i)\n",
    "variant_gene.loc[chosen]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "777"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Here is the list of GWAS genes that we are interested in\n",
    "GWAS_RA = np.unique(variant_gene.loc[chosen]['geneSymbol'])\n",
    "len(GWAS_RA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "pathways_all = pd.read_csv('PathInfo_scRNAseq.csv') \n",
    "pathways_cluster1 = pd.read_excel('TreeStructure_nodes2_scRNAseq.xlsx') "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_genes = pd.read_csv('../../../../../data/sharedData/CIA_project/DGE_data/sorted_DGEs/allTissues_sorted_expression_matrix.csv', sep = ',', index_col = 0)\n",
    "translation = pd.read_csv('../../../../../data/sharedData/CIA_project/results/orthologous_translation/orthologous_translation_NicheNet_analysis.txt', sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_DEGs = !ls '../../../../../data/sharedData/CIA_project/results/DEG_analysis/cluster_ids_fromOleg_06_29_CellType/change_mode/fdr_sorted/'\n",
    "dataset_DEGs = dataset_DEGs[4:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Number_of_genes_significant_DEGs_Sick_vs_Healthy.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_B-cells_Muscle.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_B-cells_Spleen.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Dendritic-cells_Muscle.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Dendritic-cells_Spleen.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Endothelial-cells_Joint.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Endothelial-cells_Lung.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Endothelial-cells_Muscle.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Endothelial-cells_Skin.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Erythrocytes_Muscle.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Erythrocytes_Spleen.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Fibroblasts_Joint.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Fibroblasts_Lung.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Fibroblasts_Muscle.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Fibroblasts_Skin.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Granulocytes_Joint.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Granulocytes_Muscle.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Granulocytes_Spleen.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Macrophages_Joint.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Macrophages_Lung.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Macrophages_Muscle.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Macrophages_Skin.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Macrophages_Spleen.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Monocytes_Muscle.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Monocytes_Spleen.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_NK-cells_Muscle.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_NK-cells_Spleen.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_T-cells_Joint.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_T-cells_Lung.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_T-cells_Muscle.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_T-cells_Skin.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_T-cells_Spleen.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Unknown0_Muscle.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Unknown16_Muscle.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Unknown6_Joint.csv',\n",
       " 'significant_DEGs_Sick_vs_Healthy_Unknown6_Skin.csv']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_DEGs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "DEGs_from_pathway = list()\n",
    "tissue = list()\n",
    "cell_type = list()\n",
    "DEGs_list = list()\n",
    "path = '../../../../../data/sharedData/CIA_project/results/DEG_analysis/cluster_ids_fromOleg_06_29_CellType/change_mode/fdr_sorted/'\n",
    "for subset in dataset_DEGs: #Loop through all DEG files\n",
    "    #if subset.split('_')[6].split('.')[0] == 'Muscle' or subset.split('_')[6].split('.')[0] == 'Joint':\n",
    "    #if subset.split('_')[6].split('.')[0] == 'Muscle':\n",
    "    if subset.split('_')[6].split('.')[0] == 'Joint':\n",
    "\n",
    "        subset = subset.split(' ')[-1]\n",
    "        DEGs_subset = pd.read_csv(path + subset, index_col = 0) #Read file \n",
    "        #Translate to human names\n",
    "        DEGs_subset = translation['Gene.name'][np.where(translation['Mouse.gene.name'].isin(DEGs_subset.index))[0]] \n",
    "\n",
    "\n",
    "        #Pick the shared genes between pathway and DEGs file\n",
    "        DEGs_list.append(DEGs_subset) \n",
    "\n",
    "        #to remember the tissue and cell type name\n",
    "        tissue.append(subset.split('_')[6].split('.')[0])\n",
    "        cell_type.append(subset.split('_')[5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Cluster 1\"\"\"\n",
    "pathways = pathways_all.loc[np.where(pathways_all['IngenuityCanonicalPathways'].isin\\\n",
    "                          (pathways_cluster1['IngenuityCanonicalPathways']))[0]]\n",
    "\"\"\"Cluster 2\"\"\"\n",
    "#pathways = pathways_all.loc[np.where(~pathways_all['IngenuityCanonicalPathways'].isin\\\n",
    "#                          (pathways_cluster1['IngenuityCanonicalPathways']))[0]]\n",
    "\n",
    "\n",
    "pathways.index = np.array(range(len(pathways.index)))\n",
    "\n",
    "odds_list = list()\n",
    "pval_list = list()\n",
    "for ct in np.unique(cell_type):\n",
    "    idx = np.where(np.array(cell_type) == ct)[0]\n",
    "    if len(idx) == 1:\n",
    "        DEGs_cell_type = np.array(DEGs_list[int(idx)])\n",
    "    else:\n",
    "        DEGs_cell_type = np.union1d(DEGs_list[int(idx[0])], DEGs_list[int(idx[1])])\n",
    "        \n",
    "    DEGs = list()\n",
    "    for k in range(len(pathways.index)): \n",
    "        p = pathways.loc[k]\n",
    "        DEGs.append(str(p['AllMolecules']).split(','))\n",
    "    DEGs = np.concatenate(DEGs)\n",
    "    DEGs = np.unique(DEGs)\n",
    "\n",
    "    background_genes = np.unique(DEGs_cell_type)\n",
    "    DEGs = np.intersect1d(DEGs, background_genes)\n",
    "    GWAS_RA = np.unique(GWAS_RA)\n",
    "    nonDEGs = list(set(background_genes) - set(DEGs))                \n",
    "    table11 = len(np.intersect1d(GWAS_RA, DEGs))\n",
    "    table12 = len(DEGs) - table11\n",
    "    table21 = len(np.intersect1d(GWAS_RA, nonDEGs))\n",
    "    table22 = len(background_genes) - (table21 + table12 + table11)\n",
    "    table = pd.DataFrame([[table11, table12], [table21, table22]])\n",
    "    table.index = ('DEGs', 'nonDEGs')\n",
    "    table.columns = ('GWAS', 'nonGWAS')\n",
    "    odds, pval = stats.fisher_exact(table, alternative = 'greater')\n",
    "    odds_list.append(odds)\n",
    "    pval_list.append(pval)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "GWAS_summary = pd.DataFrame((odds_list, pval_list), columns = np.unique(cell_type), index = ('odds', 'pval')).transpose()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>odds</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Endothelial-cells</th>\n",
       "      <td>3.455877</td>\n",
       "      <td>9.556580e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown6</th>\n",
       "      <td>3.070061</td>\n",
       "      <td>3.358217e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fibroblasts</th>\n",
       "      <td>3.726646</td>\n",
       "      <td>2.990161e-06</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Granulocytes</th>\n",
       "      <td>2.798690</td>\n",
       "      <td>1.950336e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Macrophages</th>\n",
       "      <td>2.387516</td>\n",
       "      <td>3.784856e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T-cells</th>\n",
       "      <td>2.084058</td>\n",
       "      <td>3.186629e-03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       odds          pval\n",
       "Endothelial-cells  3.455877  9.556580e-08\n",
       "Unknown6           3.070061  3.358217e-07\n",
       "Fibroblasts        3.726646  2.990161e-06\n",
       "Granulocytes       2.798690  1.950336e-05\n",
       "Macrophages        2.387516  3.784856e-04\n",
       "T-cells            2.084058  3.186629e-03"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GWAS_summary = GWAS_summary.sort_values('pval')\n",
    "GWAS_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "#GWAS_summary.to_csv('Joint_GWAS_summary_scRNA_cluster1_oneSided.csv', index = True, header = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "odds_list2 = list()\n",
    "pval_list2 = list()\n",
    "\n",
    "for ct in np.unique(cell_type):\n",
    "    odds_list = list()\n",
    "    pval_list = list()\n",
    "    idx = np.where(np.array(cell_type) == ct)[0]\n",
    "    if len(idx) == 1:\n",
    "        DEGs_cell_type = np.array(DEGs_list[int(idx)])\n",
    "    else:\n",
    "        DEGs_cell_type = np.union1d(DEGs_list[int(idx[0])], DEGs_list[int(idx[1])])\n",
    "\n",
    "    for j in np.unique(pathways_cluster1['subclusters']):\n",
    "        pathways_subset = pathways_cluster1.loc[np.where(pathways_cluster1['subclusters'] == j)[0]]\n",
    "        pathways_subset.index = np.array(range(len(pathways_subset.index)))\n",
    "        DEGs = list()\n",
    "        for k in range(len(pathways_subset.index)): \n",
    "            p = pathways_subset.loc[k]\n",
    "            DEGs.append(str(p['AllMolecules']).split(','))\n",
    "        DEGs = np.concatenate(DEGs)\n",
    "        DEGs = np.unique(DEGs)\n",
    "        \n",
    "        background_genes = np.unique(DEGs_cell_type)\n",
    "        DEGs = np.intersect1d(DEGs, background_genes)\n",
    "    \n",
    "        GWAS_RA = np.unique(GWAS_RA)\n",
    "        nonDEGs = list(set(background_genes) - set(DEGs))                \n",
    "        table11 = len(np.intersect1d(GWAS_RA, DEGs))\n",
    "        table12 = len(DEGs) - table11\n",
    "        table21 = len(np.intersect1d(GWAS_RA, nonDEGs))\n",
    "        table22 = len(background_genes) - (table21 + table12 + table11)\n",
    "        table = pd.DataFrame([[table11, table12], [table21, table22]])\n",
    "        table.index = ('DEGs', 'nonDEGs')\n",
    "        table.columns = ('GWAS', 'nonGWAS')\n",
    "        odds, pval = stats.fisher_exact(table, alternative = 'greater')\n",
    "        odds_list.append(odds)\n",
    "        pval_list.append(pval)\n",
    "    odds_list2.append(odds_list)\n",
    "    pval_list2.append(pval_list)\n",
    "\n",
    "# if the odds are > 1 and pval <0.05 The list of DEGs is significantly enriched by the GWAS genes\n",
    "# if the odds are < 1 and pval <0.05 The list of DEGs is not enriched by the GWAS genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "Odds = pd.DataFrame(odds_list2, index = np.unique(cell_type), columns = np.array(range(10)) + 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Endothelial-cells</th>\n",
       "      <td>3.122826</td>\n",
       "      <td>3.478670</td>\n",
       "      <td>2.856621</td>\n",
       "      <td>3.619524</td>\n",
       "      <td>3.353654</td>\n",
       "      <td>5.921626</td>\n",
       "      <td>3.922292</td>\n",
       "      <td>2.629849</td>\n",
       "      <td>3.526307</td>\n",
       "      <td>3.141188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fibroblasts</th>\n",
       "      <td>3.274154</td>\n",
       "      <td>4.644444</td>\n",
       "      <td>5.377778</td>\n",
       "      <td>3.953692</td>\n",
       "      <td>4.830508</td>\n",
       "      <td>6.300000</td>\n",
       "      <td>6.156305</td>\n",
       "      <td>1.896226</td>\n",
       "      <td>6.509255</td>\n",
       "      <td>6.504000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Granulocytes</th>\n",
       "      <td>3.102975</td>\n",
       "      <td>3.682917</td>\n",
       "      <td>2.862599</td>\n",
       "      <td>2.255319</td>\n",
       "      <td>4.086397</td>\n",
       "      <td>4.812865</td>\n",
       "      <td>3.833618</td>\n",
       "      <td>1.260522</td>\n",
       "      <td>2.860496</td>\n",
       "      <td>5.886667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Macrophages</th>\n",
       "      <td>3.545235</td>\n",
       "      <td>2.518050</td>\n",
       "      <td>3.709013</td>\n",
       "      <td>2.043735</td>\n",
       "      <td>2.734013</td>\n",
       "      <td>4.858580</td>\n",
       "      <td>3.193212</td>\n",
       "      <td>2.084507</td>\n",
       "      <td>4.204882</td>\n",
       "      <td>4.208791</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T-cells</th>\n",
       "      <td>2.262338</td>\n",
       "      <td>2.068748</td>\n",
       "      <td>1.087926</td>\n",
       "      <td>1.979545</td>\n",
       "      <td>1.254305</td>\n",
       "      <td>3.435364</td>\n",
       "      <td>2.408565</td>\n",
       "      <td>1.505017</td>\n",
       "      <td>1.991741</td>\n",
       "      <td>3.053512</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown6</th>\n",
       "      <td>3.665166</td>\n",
       "      <td>3.808571</td>\n",
       "      <td>4.905469</td>\n",
       "      <td>3.983192</td>\n",
       "      <td>5.257104</td>\n",
       "      <td>4.973844</td>\n",
       "      <td>3.704710</td>\n",
       "      <td>2.683774</td>\n",
       "      <td>3.517480</td>\n",
       "      <td>4.363454</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         1         2         3         4         5         6   \\\n",
       "Endothelial-cells  3.122826  3.478670  2.856621  3.619524  3.353654  5.921626   \n",
       "Fibroblasts        3.274154  4.644444  5.377778  3.953692  4.830508  6.300000   \n",
       "Granulocytes       3.102975  3.682917  2.862599  2.255319  4.086397  4.812865   \n",
       "Macrophages        3.545235  2.518050  3.709013  2.043735  2.734013  4.858580   \n",
       "T-cells            2.262338  2.068748  1.087926  1.979545  1.254305  3.435364   \n",
       "Unknown6           3.665166  3.808571  4.905469  3.983192  5.257104  4.973844   \n",
       "\n",
       "                         7         8         9         10  \n",
       "Endothelial-cells  3.922292  2.629849  3.526307  3.141188  \n",
       "Fibroblasts        6.156305  1.896226  6.509255  6.504000  \n",
       "Granulocytes       3.833618  1.260522  2.860496  5.886667  \n",
       "Macrophages        3.193212  2.084507  4.204882  4.208791  \n",
       "T-cells            2.408565  1.505017  1.991741  3.053512  \n",
       "Unknown6           3.704710  2.683774  3.517480  4.363454  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Odds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Odds.to_csv('Joint_Odds_summary_scRNA_oneSided.csv', index = True, header = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "Pval = pd.DataFrame(pval_list2, index = np.unique(cell_type), columns = np.array(range(10)) + 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Endothelial-cells</th>\n",
       "      <td>0.001104</td>\n",
       "      <td>3.064935e-06</td>\n",
       "      <td>6.284407e-03</td>\n",
       "      <td>0.000208</td>\n",
       "      <td>6.206997e-04</td>\n",
       "      <td>3.655343e-11</td>\n",
       "      <td>2.281222e-05</td>\n",
       "      <td>0.057082</td>\n",
       "      <td>0.000410</td>\n",
       "      <td>0.052866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Fibroblasts</th>\n",
       "      <td>0.004825</td>\n",
       "      <td>6.969792e-07</td>\n",
       "      <td>5.767664e-05</td>\n",
       "      <td>0.001606</td>\n",
       "      <td>7.036155e-05</td>\n",
       "      <td>1.471196e-08</td>\n",
       "      <td>2.521950e-07</td>\n",
       "      <td>0.236429</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.001051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Granulocytes</th>\n",
       "      <td>0.001657</td>\n",
       "      <td>3.704884e-06</td>\n",
       "      <td>1.278130e-02</td>\n",
       "      <td>0.030734</td>\n",
       "      <td>1.118395e-04</td>\n",
       "      <td>4.060871e-08</td>\n",
       "      <td>4.482997e-05</td>\n",
       "      <td>0.440616</td>\n",
       "      <td>0.006091</td>\n",
       "      <td>0.001409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Macrophages</th>\n",
       "      <td>0.000409</td>\n",
       "      <td>1.393803e-03</td>\n",
       "      <td>7.399271e-04</td>\n",
       "      <td>0.061872</td>\n",
       "      <td>8.334082e-03</td>\n",
       "      <td>7.876349e-08</td>\n",
       "      <td>9.620009e-04</td>\n",
       "      <td>0.148365</td>\n",
       "      <td>0.000096</td>\n",
       "      <td>0.011896</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T-cells</th>\n",
       "      <td>0.030413</td>\n",
       "      <td>1.572549e-02</td>\n",
       "      <td>5.134017e-01</td>\n",
       "      <td>0.069584</td>\n",
       "      <td>3.870609e-01</td>\n",
       "      <td>4.660586e-05</td>\n",
       "      <td>1.298656e-02</td>\n",
       "      <td>0.339413</td>\n",
       "      <td>0.082911</td>\n",
       "      <td>0.091669</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unknown6</th>\n",
       "      <td>0.000024</td>\n",
       "      <td>1.067897e-07</td>\n",
       "      <td>1.126156e-07</td>\n",
       "      <td>0.000006</td>\n",
       "      <td>2.157786e-08</td>\n",
       "      <td>3.638438e-10</td>\n",
       "      <td>3.764212e-06</td>\n",
       "      <td>0.016853</td>\n",
       "      <td>0.000093</td>\n",
       "      <td>0.001269</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         1             2             3         4   \\\n",
       "Endothelial-cells  0.001104  3.064935e-06  6.284407e-03  0.000208   \n",
       "Fibroblasts        0.004825  6.969792e-07  5.767664e-05  0.001606   \n",
       "Granulocytes       0.001657  3.704884e-06  1.278130e-02  0.030734   \n",
       "Macrophages        0.000409  1.393803e-03  7.399271e-04  0.061872   \n",
       "T-cells            0.030413  1.572549e-02  5.134017e-01  0.069584   \n",
       "Unknown6           0.000024  1.067897e-07  1.126156e-07  0.000006   \n",
       "\n",
       "                             5             6             7         8   \\\n",
       "Endothelial-cells  6.206997e-04  3.655343e-11  2.281222e-05  0.057082   \n",
       "Fibroblasts        7.036155e-05  1.471196e-08  2.521950e-07  0.236429   \n",
       "Granulocytes       1.118395e-04  4.060871e-08  4.482997e-05  0.440616   \n",
       "Macrophages        8.334082e-03  7.876349e-08  9.620009e-04  0.148365   \n",
       "T-cells            3.870609e-01  4.660586e-05  1.298656e-02  0.339413   \n",
       "Unknown6           2.157786e-08  3.638438e-10  3.764212e-06  0.016853   \n",
       "\n",
       "                         9         10  \n",
       "Endothelial-cells  0.000410  0.052866  \n",
       "Fibroblasts        0.000003  0.001051  \n",
       "Granulocytes       0.006091  0.001409  \n",
       "Macrophages        0.000096  0.011896  \n",
       "T-cells            0.082911  0.091669  \n",
       "Unknown6           0.000093  0.001269  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Pval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Pval.to_csv('Joint_Pval_summary_scRNA_oneSided.csv', index = True, header = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
